### PR TITLE
[GEOS-11266] csw-iso: support multiple paths mapped to queryable

### PR DESCRIPTION
**Pull request by NielsCharlier**
[GEOS-11266](https://osgeo-org.atlassian.net/browse/GEOS-11266) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geoserver&utm_campaign=badge_info)

  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [x] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

File: doc/en/user/source/extensions/csw-iso/queryables.rst [hunk 1]
```diff
@@ -0,0 +1,14 @@
+CSW ISO Metadata Profile Queryables
+===================================
+
+Mapping
+~~~~~~~
+The ISO Metadata standard (see `OGC Implementation Specification 07-045 <http://www.opengeospatial.org/standards/specifications/catalog>`_) specifies a list of so called 'queryables'. These are simple property names such as 'Title' or 'Abstract' that can be used in filters and will automatically be translated to a corresponding XPath.
+
+In some case it might be required to have an alternative mapping of queryables (to other than the default XPaths). For instance when records represent services rather than data, the corresponding XPaths are different.
```
Comment (aaime) [inline doc/en/user/source/extensions/csw-iso/queryables.rst:8]:
```suggestion
In some case it might be required to have an alternative mapping of queryables (to a different, non-default XPath). For instance when records represent services rather than data, the corresponding XPaths are different.
```

File: doc/en/user/source/extensions/csw-iso/queryables.rst [hunk 2]
```diff
@@ -0,0 +1,14 @@
+CSW ISO Metadata Profile Queryables
+===================================
+
+Mapping
+~~~~~~~
+The ISO Metadata standard (see `OGC Implementation Specification 07-045 <http://www.opengeospatial.org/standards/specifications/catalog>`_) specifies a list of so called 'queryables'. These are simple property names such as 'Title' or 'Abstract' that can be used in filters and will automatically be translated to a corresponding XPath.
```
Comment (aaime) [inline doc/en/user/source/extensions/csw-iso/queryables.rst:6]: I remember things slightly differently, "queriables" are a construct of CSW that advertises a uniform name for the queryable regardless of the underlying language being used (e.g., 'Title' should work with dublin core as well).

This is from the CSW 2.0 core spec:

image

And quoting the text:

> The goal of defining core queryable properties is query interoperability among catalogues
that implement the same protocol binding and query compatibility among catalogues that
implement different protocol bindings perhaps through the use of “bridges” or protocol
adapters. Defining a set of core queryable properties also enables simple cross-profile
discovery, where the same queries can be executed against any catalogue service without
modification and without detailed knowledge of the catalogue's information model

Long story short, a queryable mapping mechanism should be general.

File: src/extension/csw/api/src/main/java/org/geoserver/csw/store/AbstractCatalogStore.java [hunk 3]
```diff
@@ -62,28 +61,35 @@ public CloseableIterator<String> getDomain(Name typeName, final Name attributeNa
...
         // do we have such attribute?
-        final PropertyName property = rd.translateProperty(attributeName);
-        AttributeDescriptor ad = (AttributeDescriptor) property.evaluate(rd.getFeatureType());
-        if (ad == null) {
+        final List<PropertyName> properties = rd.translateProperty(attributeName);
+        boolean hasProperty = false;
+        for (PropertyName property : properties) {
+            AttributeDescriptor ad = (AttributeDescriptor) property.evaluate(rd.getFeatureType());
+            hasProperty |= ad != null;
+        }
```
Comment (aaime) [inline src/extension/csw/api/src/main/java/org/geoserver/csw/store/AbstractCatalogStore.java:69]: This could be a nice self-contained method "hasProperty"

File: src/extension/csw/api/src/main/java/org/geoserver/csw/store/AbstractCatalogStore.java [hunk 4]
```diff
@@ -62,28 +61,35 @@ public CloseableIterator<String> getDomain(Name typeName, final Name attributeNa
...
-                                                UTF_8));
+                            for (PropertyName property : properties) {
+                                Property prop = (Property) property.evaluate(feature);
+                                if (prop != null) {
+                                    values.add(
+                                            new String(
+                                                    ((String) prop.getValue()).getBytes(ISO_8859_1),
+                                                    UTF_8));
+                                    break;
+                                }
```
Comment (aaime) [inline src/extension/csw/api/src/main/java/org/geoserver/csw/store/AbstractCatalogStore.java:92]: Wondering what happens if multiple xpaths for a given property happen to exist at the same time?
The documentation should probably say something about this case.

File: src/extension/csw/csw-iso/src/main/java/org/geoserver/csw/records/iso/QueryableMappingFilterVisitor.java [hunk 5]
```diff
@@ -0,0 +1,792 @@
...
+import org.geotools.api.filter.temporal.MetBy;
+import org.geotools.api.filter.temporal.OverlappedBy;
+import org.geotools.api.filter.temporal.TContains;
+import org.geotools.api.filter.temporal.TEquals;
+import org.geotools.api.filter.temporal.TOverlaps;
+import org.geotools.data.complex.util.XPathUtil;
+import org.geotools.factory.CommonFactoryFinder;
+import org.geotools.util.factory.GeoTools;
+
+public class QueryableMappingFilterVisitor implements FilterVisitor, ExpressionVisitor {
```
Comment (aaime) [inline src/extension/csw/csw-iso/src/main/java/org/geoserver/csw/records/iso/QueryableMappingFilterVisitor.java:76]: Wow, big new class. Some javadoc explaining what it does could help.

File: doc/en/user/source/extensions/csw-iso/queryables.rst (hunk 2 above)
Comment (NielsCharlier) [reply doc/en/user/source/extensions/csw-iso/queryables.rst:6]: You are right, I didn't even know. It is a separate issue though, so I created one on jira: https://osgeo-org.atlassian.net/browse/GEOS-11323

Comment (NielsCharlier) [reply doc/en/user/source/extensions/csw-iso/queryables.rst]: I have changed the docs a bit too

**Comment by NielsCharlier** [comment]
Thanks for the review @aaime. The requested changes were made

**Comment by aaime** [comment]
Merged on main. Does it need to be backported on 2.25.x and 2.24.x?

**Comment by NielsCharlier** [comment]
Thanks, I added backports labels
